client/allocrunner/alloc_runner.go

```diff
@@ -1205,19 +1228,61 @@ func (ar *allocRunner) GetTaskEventHandler(taskName string) drivermanager.EventH
 	return nil
 }
 
-// RestartTask signalls the task runner for the  provided task to restart.
-func (ar *allocRunner) RestartTask(taskName string, taskEvent *structs.TaskEvent) error {
+// Restart satisfies the WorkloadRestarter interface and restarts all tasks
+// that are currently running.
+//
+// The event type will be set to TaskRestartRunningSignal to comply with
+// internal restart logic requirements.
+func (ar *allocRunner) Restart(ctx context.Context, event *structs.TaskEvent, failure bool) error {
+	if event.Type != structs.TaskRestartRunningSignal {
```

Comment:
Hum..not sure if I understood, do you mean something like this?

```go
func (ar *allocRunner) Restart(ctx context.Context, event *structs.TaskEvent, failure bool) error {
	ev := event.Copy()
	ev.Type = structs.TaskRestartRunningSignal
	return ar.restartTasks(ctx, ev, failure)
}
```

Reply:
I thought the original reason we were overwriting was a form of assertion? If it's not the right task restart signal, it's a bug?

Reply:
Yeah, my understanding is that, for the caller of the `allocRunner.Restart*` functions, the event type doesn't matter, so overwriting it was a way to "correct" the call if, for some reason, it was done using the wrong type.

There is no valid scenario where you would call, for example, `RestartRunning` with anything other than the `TaskRestartRunningSignal` event type, so instead of returning an error we correct the call.

The `Restart` method specifically is an interesting one. It is used to implement the `WorkloadRestarter` interface (defined [here](https://github.com/hashicorp/nomad/blob/v1.3.3/command/agent/consul/check_watcher.go#L24-L27) and [here](https://github.com/hashicorp/nomad/blob/v1.3.3/client/serviceregistration/service_registration.go#L48-L52)) so, in theory, it could be used to perform different types of restart, but I chose to keep it as-is (restart only tasks that are currently running).

Reply:
That's almost what I meant. I was suggesting we make the copy and emit both events. I was thinking that with this implementation you lose the info on which code path triggered the restart.

Reply:
A few things to note here:

- task events are used to surface information to users, they are not something like code tracing.
- event types have a fairly restrict [set of possible values](https://github.com/hashicorp/nomad/blob/v1.3.3/nomad/structs/structs.go#L8005-L8098), so they are not the best to help trace code path.
- events have other, more meaningful fields, like `DisplayMessage`, `Details`, `ReastartReason` etc. and those will be kept from the input.
- task events are stored in state, and we only keep the last 10, so emitting (almost) duplicate events will take up a slot and provide little value to users.

That being said, my implementation does abuse the task event mechanism a bit. I tried to think of an alternative, but couldn't think of one yet. I will take another look to see if I can think of something else.

Reply:
https://github.com/hashicorp/nomad/pull/14127/commits/e620cd187452b45ab04bf65d56aa2f90fabfc9eb removes all of this task event nonsense 😅 

I also got ride of the new task event types to avoid any confusion on the developer side as to which type to use. I think it's easy enough for users to distinguish between them given their [event description](https://github.com/hashicorp/nomad/blob/e620cd187452b45ab04bf65d56aa2f90fabfc9eb/client/client.go#L939-L953):
![image](https://user-images.githubusercontent.com/775380/186511121-56002e71-b280-4655-9144-938b2efd6caa.png)